Core/include/Acts/Navigation/NextNavigator.hpp

```diff
@@ -0,0 +1,357 @@
+
+        ACTS_VERBOSE("current portal set to "
+                     << nState.currentPortal->surface().geometryId());
+      } else {
+        ACTS_VERBOSE("this is a surface, storing it.");
+
+        // If we are on the surface pointed at by the iterator, we can make
+        // it the current one to pass it to the other actors
+        nState.currentSurface = nextSurface;
+        ACTS_VERBOSE("current surface set to "
+                     << nState.currentSurface->geometryId());
+        ++nState.surfaceCandidate;
+      }
+    }
+  }
+
+  /// @brief Navigator target call
+  ///
+  /// @tparam propagator_state_t is the type of Propagatgor state
+  /// @tparam stepper_t is the used type of the Stepper by the Propagator
+  ///
+  /// @param [in,out] state is the mutable propagator state object
+  /// @param [in] stepper Stepper in use
+  template <typename propagator_state_t, typename stepper_t>
+  void target(propagator_state_t& state, const stepper_t& stepper) const {
+    ACTS_VERBOSE(volInfo(state)
+                 << posInfo(state, stepper) << "Entering navigator::target.");
+
+    auto& nState = state.navigation;
+
+    if (inactive()) {
+      ACTS_VERBOSE("navigator inactive");
+      return;
+    }
+
+    if (nState.currentVolume == nullptr) {
+      initializeTarget(state, stepper);
+    }
+
+    if (nState.currentSurface != nullptr) {
+      ACTS_VERBOSE("stepping through surface");
+    } else if (nState.currentPortal != nullptr) {
+      ACTS_VERBOSE("stepping through portal");
+
+      nState.surfaceCandidates.clear();
+      nState.surfaceCandidate = nState.surfaceCandidates.end();
+
+      nState.currentPortal->updateDetectorVolume(state.geoContext, nState);
+
+      initializeTarget(state, stepper);
+    }
+
+    for (; nState.surfaceCandidate != nState.surfaceCandidates.end();
+         ++nState.surfaceCandidate) {
+      // Screen output how much is left to try
+      ACTS_VERBOSE(volInfo(state)
+                   << std::distance(nState.surfaceCandidate,
+                                    nState.surfaceCandidates.end())
+                   << " out of " << nState.surfaceCandidates.size()
+                   << " surfaces remain to try.");
+      // Take the surface
+      const auto& c = *(nState.surfaceCandidate);
+      const auto& surface =
+          (c.surface != nullptr) ? (*c.surface) : (c.portal->surface());
+      // Screen output which surface you are on
+      ACTS_VERBOSE("Next surface candidate will be " << surface.geometryId());
+      // Estimate the surface status
+      bool boundaryCheck = c.boundaryCheck;
+      auto surfaceStatus = stepper.updateSurfaceStatus(state.stepping, surface,
+                                                       boundaryCheck, logger());
+      if (surfaceStatus == Intersection3D::Status::reachable) {
+        ACTS_VERBOSE("Surface reachable, step size updated to "
+                     << stepper.outputStepSize(state.stepping));
+        break;
+      }
+    }
+
```

Comment:
Add a comment saying that in the target, we invalidate both current surface and portal, because we do not want to fall into loops.